include/trieste/intrusive_ptr.h

```diff
@@ -0,0 +1,350 @@
+    };
+  }
+
+  template<typename T>
+  struct intrusive_refcounted_traits
+  {
+    static constexpr void intrusive_inc_ref(T* ptr)
+    {
+      ptr->intrusive_inc_ref();
+    }
+
+    static constexpr void intrusive_dec_ref(T* ptr)
+    {
+      ptr->intrusive_dec_ref();
+    }
+  };
+
+  template<typename T>
+  struct intrusive_ptr final
+  {
+  private:
+    T* ptr;
+
+    constexpr void inc_ref()
+    {
+      if (ptr)
+      {
+        intrusive_refcounted_traits<T>::intrusive_inc_ref(ptr);
+      }
+    }
+
+    constexpr void dec_ref()
+    {
+      if (ptr)
+      {
+        intrusive_refcounted_traits<T>::intrusive_dec_ref(ptr);
+        ptr = nullptr;
+      }
+    }
+
+  public:
+    template<typename... Args>
+    static intrusive_ptr<T> make(Args&&... args)
+    {
+      return intrusive_ptr(new T(std::forward<Args>(args)...));
+    }
+
+    constexpr intrusive_ptr() : ptr{nullptr} {}
+
+    constexpr intrusive_ptr(std::nullptr_t) : ptr{nullptr} {}
+
+    constexpr explicit intrusive_ptr(T* ptr_) : ptr{ptr_}
+    {
+      inc_ref();
+    }
+
+    template<typename U>
+    constexpr intrusive_ptr(const intrusive_ptr<U>& other) : ptr{other.ptr}
+    {
+      inc_ref();
+    }
+
+    template<typename U>
+    constexpr intrusive_ptr(intrusive_ptr<U>&& other) : ptr{other.release()}
+    {}
+
+    constexpr intrusive_ptr(const intrusive_ptr<T>& other) : ptr{other.ptr}
+    {
+      inc_ref();
+    }
+
+    constexpr intrusive_ptr(intrusive_ptr<T>&& other) : ptr{other.release()} {}
+
+    constexpr intrusive_ptr<T>& operator=(const intrusive_ptr<T>& other)
+    {
+      // Self-assignment case, don't bother touching refcounts then
+      if (ptr == other.ptr)
+      {
+        return *this;
+      }
+
+      intrusive_ptr<T> tmp;
+      // Don't actually inc_ref, but putting old ptr in tmp lets us leverage the
+      // built in dec_ref with null checks below.
+      tmp.ptr = ptr;
+      ptr = other.ptr;
+      inc_ref();
+      // tmp gets dec_ref here, potentially destroying the value at old ptr
+      return *this;
+    }
+
+    constexpr intrusive_ptr<T>& operator=(intrusive_ptr<T>&& other)
+    {
+      std::swap(ptr, other.ptr);
+      return *this;
+    }
+
+    constexpr void swap(intrusive_ptr<T>& other)
+    {
+      std::swap(ptr, other.ptr);
+    }
+
+    constexpr void reset()
+    {
+      dec_ref();
+    }
+
+    constexpr T* get() const
+    {
+      return ptr;
+    }
+
+    constexpr T* operator->() const
+    {
+      return get();
+    }
+
+    constexpr T& operator*() const
+    {
+      return *get();
+    }
+
+    constexpr operator bool() const
+    {
+      return ptr;
+    }
+
+    constexpr T* release()
+    {
+      auto p = get();
+      ptr = nullptr;
+      return p;
+    }
+
+    ~intrusive_ptr()
+    {
+      dec_ref();
+    }
+
+    friend std::hash<intrusive_ptr<T>>;
+  };
+
+  template<typename T>
+  struct intrusive_refcounted
+  {
+  private:
+    detail::copyable_refcount intrusive_refcount = 0;
```

Comment:
The assignment can end up calling a copy constructor, it is better to use the initialisation syntax `{foo}` rather than `= foo`.
```suggestion
    detail::copyable_refcount intrusive_refcount{1};
```
Also, a more common pattern is to start with `1`, and then you don't need an incref on construction.  The incref should could get optimised, but if it is using atomics, then it is probably not allowed to optimise.

Reply:
Ok, so my existing design comes from the idea that we should support "ingesting" a raw pointer into an intrusive_ptr, whether it's owned or not.

If we start at refcount 1 instead, then we have to know if we're on the first construction (in which case no increment), or if we're taking joint ownership (in which case, do increment) (yes via a raw pointer, yes rare).

This is not a problem for other refcounts where we control the initialization more. For example, shared_ptr completely hides its control block's lifecycle, so it can probably just start at 1 with no problem. Our control block is just the object itself, and we let users allocate one whenever they want.

The trouble is, if we ever write `intrusive_ptr{new X}`, and X's constructor started its refcount at 1, then we might end up getting a refcount of 2. Conversely, if we assume the raw ptr construction must have a refcount of 1 (though that is assertable I suppose), then it's possible to mess up and not increment in some cases (which admittedly are usually in utility methods like `intrusive_ptr_from_this` or intrusive_ptr casting, but it is complexity nonetheless).

For context, there are unavoidable `intrusive_ptr{new NodeDef}` instances that are needed because the NodeDef constructor is private and `make` can't call it.

So basically I can implement start-at-1, but it will have some code complexity cost. What do you think?

Reply:
Resolved via voice call:
- keep starting value at 0
- comment why, summarizing this conversation
- make sure the 0 only shows up once (likely, remove copyable_refcount's default constructor)